cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeMDS.java

```diff
@@ -32,91 +36,117 @@
+      MDSKey.Splitter splitter = entry.getKey().split();
+      // first part is TYPE_TO_MODULE_PREFIX
+      splitter.skipString();
+      // second part is namespace
+      splitter.skipString();
+      // third part is the type name, which is what we expect
+      String key = splitter.getString();
+      types.add(getTypeMeta(namespaceId, key, entry.getValue()));
     }
     return types;
   }
 
-  public void write(DatasetModuleMeta moduleMeta) {
-    put(getModuleKey(moduleMeta.getName()), moduleMeta);
+  public void writeModule(Id.Namespace namespaceId, DatasetModuleMeta moduleMeta) {
+    Id.DatasetModule datasetModuleId = Id.DatasetModule.from(namespaceId, moduleMeta.getName());
+    write(getModuleKey(namespaceId.getId(), moduleMeta.getName()), moduleMeta);
 
     for (String type : moduleMeta.getTypes()) {
-      write(type, moduleMeta.getName());
+      writeTypeToModuleMapping(Id.DatasetType.from(namespaceId, type), datasetModuleId);
     }
   }
 
-  public void deleteModule(String name) {
-    DatasetModuleMeta module = getModule(name);
+  public void deleteModule(Id.DatasetModule datasetModuleId) {
+    DatasetModuleMeta module = getModule(datasetModuleId);
     if (module == null) {
       // that's fine: module is not there
       return;
     }
 
-    delete(getModuleKey(module.getName()));
+    deleteAll(getModuleKey(datasetModuleId.getNamespaceId(), datasetModuleId.getId()));
 
     for (String type : module.getTypes()) {
-      delete(getTypeKey(type));
+      deleteAll(getTypeKey(datasetModuleId.getNamespaceId(), type));
+    }
+  }
+
+  public void deleteModules(Id.Namespace namespaceId) {
+    Collection<DatasetModuleMeta> modules = getModules(namespaceId);
+    for (DatasetModuleMeta module : modules) {
+      deleteModule(Id.DatasetModule.from(namespaceId, module.getName()));
     }
   }
 
-  private DatasetTypeMeta getTypeMeta(String typeName, String moduleName) {
-    DatasetModuleMeta moduleMeta = getModule(moduleName);
-    return getTypeMeta(typeName, moduleMeta);
+  private DatasetTypeMeta getTypeMeta(Id.Namespace namespaceId, String typeName, Id.DatasetModule datasetModuleId) {
+    DatasetModuleMeta moduleMeta = getModule(datasetModuleId);
+    return getTypeMeta(namespaceId, typeName, moduleMeta);
   }
 
-  private DatasetTypeMeta getTypeMeta(String typeName, DatasetModuleMeta moduleMeta) {
+  private DatasetTypeMeta getTypeMeta(Id.Namespace namespaceId, String typeName, DatasetModuleMeta moduleMeta) {
     List<DatasetModuleMeta> modulesToLoad = Lists.newArrayList();
     // adding first all modules we depend on, then myself
     for (String usedModule : moduleMeta.getUsesModules()) {
-      modulesToLoad.add(getModule(usedModule));
+      // Try to find module in the specified namespace first
+      DatasetModuleMeta usedModuleMeta = getModule(Id.DatasetModule.from(namespaceId, usedModule));
+      // if not found, try to load it from system namespace
```

Comment:
Yes, should try to consolidate. I'll see what I can do. The problem I think is the error handling is different in each case, but I'll take another stab at it.